src/NumberTheory/QuadFormAndIsom/lattices_with_isometry.jl

```diff
@@ -1596,74 +1597,87 @@ julia> order(G)
+function image_centralizer_in_Oq(Lf::ZZLatWithIsom; _local::Bool=false)
   @req is_integral(Lf) "Underlying lattice must be integral"
   n = order_of_isometry(Lf)
   L = lattice(Lf)
   f = isometry(Lf)
   chi = minpoly(Lf)
-  if is_unimodular(L)
-    # If L is unimodular we do not have to do any wizard's trick since the
-    # discriminant group is trivial
-    qL = discriminant_group(L)
-    OqL = orthogonal_group(qL)
-    return OqL, id_hom(OqL)
-  elseif (n in [1, -1]) || (isometry(Lf) == -identity_matrix(QQ, rank(L)))
-    # Trivial cases: the lattice has rank 0 or 1, or it is endowed with +- identity
-    return image_in_Oq(L)
-  elseif rank(L) == degree(chi)
-    # Hermitian type with hermitian structure of rank 1
-    # The image consists of -id and multiplication by a primitive element on the
-    # hermitian side, which itself corresponds to f along the trace construction
-    qL, fqL = discriminant_group(Lf)
-    OqL = orthogonal_group(qL)
-    UL = elem_type(OqL)[OqL(m; check=false) for m in ZZMatrix[-matrix(one(OqL)), matrix(fqL)]]
-    return sub(OqL, unique!(UL))
-  elseif is_of_hermitian_type(Lf)
-    if is_definite(L) || rank(L) == 2
-      # If L is definite or of rank 2 and we use the correspondence between
-      # O(L, f) and O(L, h) via the trace equivalence, where (L, h) is the
-      # hermitian structure on (L, f). Preferable choice than computing O(L, f)
-      # directly because the rank of the hermitian structure is a strict divisor
-      # of the rank of L (the lower the rank is, the better!)
-      H, res = hermitian_structure_with_transfer_data(integer_lattice(; gram=gram_matrix(L)), f)
-      @hassert :ZZLatWithIsom 1 is_definite(H) # Should always be the case since L is definite, even for infinite isometries
-      OH = isometry_group(H)                   # This is identified with O(L, f) using `res`
-      geneOH = gens(OH)
-      geneUL = QQMatrix[_transfer_isometry(res, matrix(g)) for g in geneOH]
-      @hassert :ZZLatWithIsom 1 all(g -> g*gram_matrix(Lf)*transpose(g) == gram_matrix(Lf), geneUL)
-      @hassert :ZZLatWithIsom 1 all(g -> g*f == f*g, geneUL)
-      UL = matrix_group(unique!(geneUL))
-      disc = discriminant_representation(L, UL; check=false, ambient_representation=false)
-      return image(disc)
+  T = Tuple{AutomorphismGroup{TorQuadModule}, GAPGroupHomomorphism{AutomorphismGroup{TorQuadModule}, AutomorphismGroup{TorQuadModule}}}
+  
+  if _local 
+    return get_attribute!(Lf, :image_centralizer_in_Oq_local) do
+      qL,fqL = discriminant_group(Lf)
+      OqL = orthogonal_group(qL)
+      C = centralizer(OqL, OqL(matrix(fqL)))
+      return C::T
+    end
+  end
+  
+  return get_attribute!(Lf, :image_centralizer_in_Oq) do
+    if is_unimodular(L)
+      # If L is unimodular we do not have to do any wizard's trick since the
+      # discriminant group is trivial
+      qL = discriminant_group(L)
+      OqL = orthogonal_group(qL)
+      return (OqL, id_hom(OqL))::T
```

Comment:
Same problem: you need to annotate the value returned by `get_attribute!` -- the annotation here is useless as it is "eaten up" by `get_attribute!`
```suggestion
      return (OqL, id_hom(OqL))
```

Reply:
Thanks!